src/common/main.lua

```diff
@@ -1,28 +1,143 @@
 local json = require(".common.json")
 local utils = require(".common.utils")
 local main = {}
+-- just to ignore lint warnings
+local ao = ao or {}
 
 local camel = utils.camelCase
+
+local function errorHandler(err)
+	return debug.traceback(err)
+end
+
 main.init = function()
-	Owner = Owner or ao.env.Process.Owner
-	Name = Name or "ar.io ANT Registry"
+	ANTS = ANTS or {}
+	-- Example ANT structure
+	-- ANTS["antId"] = {
+	--     Owner = "userId",
+	--     Controllers = {"userId1", "userId2"},
+	-- }
+	USERS = USERS or {}
+	-- Example USERS structure
+	-- USERS["userId"] = {"antId1", "antId2"}
 
 	local ActionMap = {
 		Register = "Register",
 		StateNotice = "State-Notice",
 		AccessControlList = "Access-Control-List",
 	}
 
-	Handlers.add("info", Handlers.utils.hasMatchingTag("Action", "Info"), function(msg)
-		ao.send({
-			Target = msg.From,
-			Name = Name,
-			Owner = Owner,
-			Data = json.encode({
-				Name = Name,
-				Owner = Owner,
-			}),
-		})
+	Handlers.add(camel(ActionMap.Register), Handlers.utils.hasMatchingTag("Action", ActionMap.Register), function(msg)
+		print("Action: " .. ActionMap.Register)
+
+		local antId = msg.Tags["Process-Id"]
+		assert(type(antId) == "string", "Process-Id tag is required")
+
+		if ANTS[antId] then
+			ao.send({
+				Target = msg.From,
+				Action = "Register-Notice-Failure",
+				["Message-Id"] = msg.Id,
+				Data = "ANT is already registered",
+			})
+		else
+			ANTS[antId] = {
+				Owner = nil,
+				Controllers = {},
+			}
+
+			ao.send({
+				Target = antId,
+				Action = "State",
+			})
+			ao.send({
+				Target = msg.From,
+				Action = "Register-Notice",
+			})
+		end
+	end)
+
+	Handlers.add(
+		camel(ActionMap.StateNotice),
+		Handlers.utils.hasMatchingTag("Action", ActionMap.StateNotice),
+		function(msg)
+			print("Action: " .. ActionMap.StateNotice)
+
+			local stateStatus, stateRes = xpcall(function()
+				return utils.parseAntState(msg.Data)
+			end, errorHandler)
+			if not stateStatus then
+				ao.send({
```

Comment:
What's the presumed utility of responding in this way?

Reply:
Its a pattern we implemented in other processes - we use "Error" tag and "Action" tag in the ar.io sdk for example so that we can throw an error on the client side and drive workflows. @dtfiedler might have more to say on it.